### PR TITLE
Log download filtering

### PR DESCRIPTION
**Pull request by peterbarker**
This has been hanging around for a long time.  Thought I would PR it.

### CODE REVIEW

File: libraries/DataFlash/DataFlash_File.cpp [hunk 1]
```diff
@@ -451,6 +451,15 @@ char *DataFlash_File::_log_file_name_long(const uint16_t log_num) const
     return buf;
 }
 
+char *DataFlash_File::_log_filtered_file_name(const uint16_t log_num) const
+{
+    char *buf = nullptr;
+    if (asprintf(&buf, "%s/%u.FIN", _log_directory, (unsigned)log_num) == 0) {
```
Comment (magicrub) [inline libraries/DataFlash/DataFlash_File.cpp]: .FIN?

**Comment by WickedShell** [comment]
Since I don't see a PR for the MAVLink stuff seperately, will just tag it here.

```
+    <message id="12002" name="LOG_FILTER_ADD">
+      <description>Create a filter to use for subsequent log downloads.</description>
+      <field type="uint8_t" name="target_system">System ID</field>
+      <field type="uint8_t" name="target_component">Component ID</field>
+      <field enum="LOG_FILTER_CREATE_FLAGS" name="axis" type="uint8_t"></field>
+      <field name="seq" type="uint8_t">Filter fields sequence number</field>
+      <field type="char[128]" name="fields">comma-separated field names</field>
+    </message>
```

I can't find an actual enum for LOG_FILTER_CREATE_FLAGS, intended usage of `seq` is unclear. How does a GCS know if the filters were uploaded correctly?

**Comment by peterbarker** [comment]
On Mon, 18 Jun 2018, Tom Pittenger wrote:

> +char *DataFlash_File::_log_filtered_file_name(const uint16_t log_num) const
> +{
> +    char *buf = nullptr;
> +    if (asprintf(&buf, "%s/%u.FIN", _log_directory, (unsigned)log_num) == 0
> ) {
> 
> .FIN?

Jaws reference.

Or, possibly Filtered bINary file.

I forget which.

**Comment by peterbarker** [comment]
On Mon, 18 Jun 2018, WickedShell wrote:

> +    <message id="12002" name="LOG_FILTER_ADD">
> +      <description>Create a filter to use for subsequent log downloads.</de
> scription>
> +      <field type="uint8_t" name="target_system">System ID</field>
> +      <field type="uint8_t" name="target_component">Component ID</field>
> +      <field enum="LOG_FILTER_CREATE_FLAGS" name="axis" type="uint8_t"></fi
> eld>
> +      <field name="seq" type="uint8_t">Filter fields sequence number</field
> >
> +      <field type="char[128]" name="fields">comma-separated field names</fi
> eld>
> +    </message>
> 
> I can't find an actual enum for LOG_FILTER_CREATE_FLAGS, intended usage of
> seq is unclear. How does a GCS know if the filters were uploaded correctly?

The seq would be there in case your "fields" field exceeded 128 bytes.  We 
could eliminate that if we just provided feedback to the GCS.

Not sure what I was thinking with the flags.  Perhaps we could do away 
with the CLEAR message in favour of a "CLEAR" flag?

Checking if the filter is correct.... not sure how we would do that, but 
important to add.  Possibly an additional message?


This was really just a PoC done when @WickedShell mentioned it some time 
ago.

It did actually work end-to-end at the time; the MAVProxy branch is here:
https://github.com/ArduPilot/MAVProxy/compare/master...peterbarker:log-download-filtering?expand=1

(ignore the waypoint stuff - no idea if that's correct any more!)

**Comment by peterbarker** [comment]
The vehicle shouldn't care about whether their set is complete.  That's for the GCS to ensure.

What's missing is a message that can be requested by the GCS that indicates what the current filtering is set to.  And if the GCS doesn't like what it sees it updates them.  Does that seem a reasonable way forward?  (pre-message-interval-patch we might just spew out the current filter flags @1Hz?  OTOH, this is unlikely to go in before message-interval)

**Comment by peterbarker** [comment]
I've added a patch to the mavlink repo branch that adds:

```
    <message id="12003" name="LOG_FILTER">
      <description>Announce filter currently being applied to downloaded log files.</description>
      <field name="seq_count" type="uint8_t">Number of these messages required to assemble complete set</field>
      <field name="seq" type="uint8_t">Filter fields sequence number, 0..(seq_count-1)</field>
      <field type="char[128]" name="fields">comma-separated field names</field>
    </message>
```

**Comment by WickedShell** [comment]
Why not pass a single tag name per message up for filters? I know it's annoying for message overhead on the GCS, but you don't do this very often, and if you can make the assumption of 1 per, then the GCS getting the filter can be a request by index, and the vehicle can easily find the information, (with a final terminator that says "this value is not set" which lets the GCS know when to stop requesting). I'm just concerned that the mangling of the `fields` entry on either side will be poor when you try to accept a lot of stuff through the filter. I'll have to look at the actual impl code you provided here later though. For now my rough (tired) thought on messages is:

```
    <message id="12003" name="LOG_FILTER_REQ">
      <description>Request for a log filter entry.</description>
      <field name="seq" type="uint16_t">Sequence number of the filter</field>
      <field type="uint8_t" name=target_system" />
      <field type="uint8_t" name=target_component" />
    </message>
    <message id="12003" name="LOG_FILTER">
      <description>Request for a log filter entry.</description>
      <field name="seq" type="uint16_t">Sequence number of the filter</field>
      <field type="uint8_t" name=target_system" />
      <field type="uint8_t" name=target_component" />
      <field type="char[8]" name="field">Field name to be filtered, null string for no entry</field>
    </message>
```

Not sure if that rambling made any sense, will try and come back to this once I'm more awake.

**Comment by WickedShell** [comment]
#3187 has the original issue that prompted this, as well as performance numbers as of when the patch was first done.

**Comment by peterbarker** [comment]
Insufficient interest in this feature to pursue it.

**Pull request closed without merging**